src/class/Socket/Socket.cpp

```diff
@@ -0,0 +1,145 @@
+
+#include <Socket/Socket.hpp>
+
+/*
+** ------------------------------- CONSTRUCTOR --------------------------------
+*/
+
+Socket::Socket(IP ip, Port port)
+	: _ip(ip)
+	, _port(port)
+	, _binded(false)
+	, _fd(-1)
+	, _domain(AF_INET)
+	, _type(SOCK_STREAM)
+	, _protocol(0)
+{
+	/* Create socket */
+	this->_fd = socket(this->_domain, this->_type, this->_protocol);
+	if (this->_fd <= -1)
+		throw SocketCreationFailed();
+
+	/* Set the socket to reuse (makes it re-bindable directly after close) */
+	int reuse_addr = true;
+	setsockopt(this->_fd, SOL_SOCKET, SO_REUSEADDR, &reuse_addr, sizeof(reuse_addr));
+
+	/* Init sockaddr */
+	memset(&this->_sockaddr, 0, sizeof(this->_sockaddr));
+	this->_sockaddr.sin_family = this->_domain;
+	this->_sockaddr.sin_port = htons(this->_port.get());
+	this->_sockaddr.sin_addr.s_addr = inet_addr(this->_ip.get().c_str());
+}
+
+Socket::Socket(Socket const &src)
+	: _ip(src._ip)
+	, _port(src._port)
+	, _binded(src._binded)
+	, _fd(src._fd)
+	, _domain(src._domain)
+	, _type(src._type)
+	, _protocol(src._protocol)
+	, _sockaddr(src._sockaddr)
+{
+}
+
+const char *Socket::SocketCreationFailed::what() const throw()
+{
+	return "Socket creation failed";
+}
+const char *Socket::SocketListeningFailed::what() const throw()
+{
+	return "Socket setting to listening mode failed";
+}
+const char *Socket::SocketBindingFailed::what() const throw()
+{
+	return "Socket binding failed";
+}
+const char *Socket::SocketNotBinded::what() const throw()
+{
+	return "Socket not binded";
+}
+
+/*
+** -------------------------------- DESTRUCTOR --------------------------------
+*/
+
+Socket::~Socket()
+{
+}
+
+/*
+** --------------------------------- OVERLOAD ---------------------------------
+*/
+
+std::ostream &operator<<(std::ostream &o, Socket const &i)
+{
+	o << "Socket[" << i.getFd() << "]";
+	return o;
+}
+
+/*
+** --------------------------------- METHODS ----------------------------------
+*/
+
+void Socket::bind(void)
+{
+	int bind_status;
+	if (this->_binded)
+		return;
+	bind_status = ::bind(this->_fd, reinterpret_cast<const sockaddr *>(&this->_sockaddr), sizeof(this->_sockaddr));
+	if (bind_status <= -1)
+		throw SocketBindingFailed();
+	this->_binded = true;
+}
+
+void Socket::listen(void)
+{
+	int listen_status;
+	if (!this->_binded)
+		throw SocketNotBinded();
+	listen_status = ::listen(this->_fd, MAX_SYS_BACKLOG);
```

Comment:
about, MAX_SYS_BACKLOG.
if we reach this amount of clients calling this socket and the queue is full then every new connection will fail automatically?
if so, why you are using 128?
and is there a way for us to know that the queue is full?

Reply:
I have been reading about this and there is just a max ammount of connections in the queue, the limit of the queue depends on the system and you can check it using a command (which I dont remember now) but my system max was 128, thats why I used that number, I'll take a deeper look into which number we should use